63185

wolfebrian2120 opened the pull request:
## Description
<!--- Provide a general summary of your changes in the Title above -->
Completed the LDAP portion of 63185. Basically added a checkbox to the Extended LDAP Request form that allows a SSL/TLS connection to trust any server certificate that is presented to the jmeter client.

<!--- Describe your changes in detail here -->
I added the checkbox to the config and sampler forms.
I added a custom Socket Class that uses a custom TrustManager to replace the default trust manager used by the jndi connection.
Updated the connect methods to take in  a new parameter and if that parameter is set it switches the trustmanager for that connection
I updated the messages.properties and their respective languages that have ldap_secure defined with another parameter called ldap_trust_all. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
As the Enhancement states having the option to trust all SSL/TLS connections to ldap servers improves testing and development of ldap load tests. While the connection is less secure it provides the option to the script writer to decide that. Default behaviour is to use the standard trust manager and provides all of the normal validation of the JVM.
This change helps improve productivity and use for the end user that just wants to connect to an LDAPS port and start testing. It allows the script writer the flexibility to get something up and running more quickly without having to worry about importing or setting up certificate trusts for the jmeter JVM.
<!--- If it fixes an open issue, please link to the issue here. -->
Issue 63185
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
I created a jmeter script with Extended LDAP Sampler that did a thread bind. I configured it to connect to an Oracle Unified Directory instance over LDAPS. Ensured that Secure Protocol and Trust All check boxes were checked.

I ran this test in a lab environment and also ran a couple other distributed tests against a production environment. This included custom scripts that would loop and perform logic before executing the sampler.

These tests included connecting to the host using different DNS names than what was on the given certificate, which would result in an exception being thrown about DNS name not matching a subject alternative name in the certificate.

Other cases that should be handled are certificate path exceptions. If a server is using a non standard CA and it hasn't been imported into the JVM cacerts keystore.

<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
This change only affects the Extended LDAP Request samplers and the config objects for that sampler set. The basic ldap sampler is not affected by this change.
This change adds in a newly implemented socket class that does the necessary certificate validations or the no longer necessary cert validations, lol.
## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Delete as appropriate -->
- New feature (non-breaking change which adds functionality)
Merely adds a checkbox and an additional option regarding the LDAPS client connections coming from a jmeter script

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X ] My code follows the [code style][style-guide] of this project.
I ran the code changes against the code style ant checks. I made all the necessary changes to make it pass those checks.
- [X ] I have updated the documentation accordingly.
documentation does not cover setting up SSL/TLS connections for LDAP samplers, so I did not update any of that documentation. However use of the LDAP sampler is rather straightforward.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

bwolfehcm committed:
added ldap trustall functionality. built custom trust manager and added checkbox to activate it on the sampler configuration.

bwolfehcm committed:
Updated Language messages.

Also ran JUnit tests using the following ant command
ant -Ddisable-svnCheck=true -Djava.awt.headless=true -Drmi_force_localhost=true -Dskip.bug52310=true test

bwolfehcm committed:
removed LDAP-test.jmx

pmouawad commented on docs/usermanual/build-ldapext-test-plan.html:
```diff
@@ -442,7 +442,10 @@ <h2 id="add_threadbind">8b.3.1 Adding a Thread bind Request<a class="sectionlink
 
 <i>(Optional)</i> Check the box Use Secure LDAP Protocol if you access with LDAP over SSL (ldaps)
 </li>
+<li>
```
For information, you don't need to commit HTML, it's generated from xml in xdocs

pmouawad commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/TrustAllSocket.java:
```diff
@@ -0,0 +1,121 @@
+            @Override
+            public void checkClientTrusted(X509Certificate[] arg0, String arg1, Socket arg2)
+                throws CertificateException {
+                // TODO Auto-generated method stub
```
Can you remove those TODO ? Thanks

pmouawad commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/TrustAllSocket.java:
```diff
@@ -0,0 +1,121 @@
+            ctx = SSLContext.getInstance("TLS");
+        } catch (NoSuchAlgorithmException e1) {
+            // TODO Auto-generated catch block
+            e1.printStackTrace();
```
Logging is better than printStackTrace(). IMO let it throw.

bwolfehcm committed:
removed html file from docs directory related to the ldap ext sampler.
removed to dos from TrustAllSocket.java
removed printStackTrace and threw exceptions. also added handling where the instance is created in getDefault. printing localized messages at error level in log.

bwolfehcm committed:
updated TrustAllSocket.java to comply with the style checker

wolfebrian2120 commented:
Ok I have made all of the suggested changes.

FSchumacher commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/LDAPExtSampler.java:
```diff
@@ -85,6 +85,8 @@
 
     public static final String SECURE = "secure"; // $NON-NLS-1$
+    
+    public static final String TRUSTALL = "TrustAll";
```
The other key values are all lowercase. Any special reason for CamelCase?

wolfebrian2120 replied:
no, just working with command line tools that have that explicit option have put me in the habit of spelling it that way. I will update it.

FSchumacher commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/LDAPExtSampler.java:
```diff
@@ -660,7 +670,7 @@ private void bindOp(SampleResult res) throws NamingException {
         try {
             res.sampleStart();
-            ctx = LdapExtClient.connect(getServername(), getPort(), getRootdn(), getUserDN(), getUserPw(),getConnTimeOut(),isSecure());
+            ctx = LdapExtClient.connect(getServername(), getPort(), getRootdn(), getUserDN(), getUserPw(),getConnTimeOut(),isSecure(),isTrustAll());
```
Again spacepolice time :) plus you could add a line break to shorten the lines a bit.

wolfebrian2120 replied:
yup will do.

FSchumacher commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/LdapExtClient.java:
```diff
@@ -78,14 +78,19 @@ public LdapExtClient() {
             sb.append("ldaps://"); // $NON-NLS-1$
+            if (trustAll){
+                log.debug("Secure true, trustAll true");
+                env.put("java.naming.ldap.factory.socket", "org.apache.jmeter.protocol.ldap.sampler.TrustAllSocket");
```
My initial thought was to add a parameter that would have been inserted here, but maybe the switch for `trustAll` is a bit more user friendly.

wolfebrian2120 replied:
Yeah I went with that because I wanted the user to be able to pick whether they wanted to use the new feature or not.

FSchumacher commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/TrustAllSocket.java:
```diff
@@ -0,0 +1,121 @@
+            ctx.init(null, trustAllCerts, new java.security.SecureRandom());
+        } catch (KeyManagementException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
```
Use a logger to show error messages.

wolfebrian2120 replied:
nice catch. I thought i got them all

FSchumacher reviewed (changes requested):
Nice work.

bwolfehcm committed:
Added spacing and line breaks for code readability. Also updated TRUSTALL string in LDAPExtSamler to conform with all other strings

ham1 commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/TrustAllSocket.java:
```diff
@@ -0,0 +1,117 @@
+            try {
+                defaultFactory.compareAndSet(null, new TrustAllSocket());
+            } catch (KeyManagementException e) {
+                log.error("KeyManagementException: "+e.getLocalizedMessage());
```
Would passing the exception into the logger be best, and maybe a more descriptive error message?
```suggestion
                log.error("KeyManagementException while trying to ... ", e);
```

wolfebrian2120 replied:
honestly I don't think this error will ever happen as the custom trust manager doesn't do anything that will generate the error. Its just part of the interface. The only way the NoSuchAlgorithmException will be thrown is if the JDK does not support some version of TLS. The KeyManagementException would only be thrown in the event there is an error dealing initializing the custom SSLContext object. which is only dependent on the inputs from the custom trust manager which basically does nothing and the SecureRandom() function, which should always return some value. Both of which would be an indication of a broken JDK/JRE or a really old version.
I suppose I could provide a nicer message indicating some context of what we are trying to do here in the event an error occurs and someone sends the output to us for analysis. 
I guess something along the lines of "KeyManagementException when changing SSLContext for TrustAll" or something like that. That may help someone who isn't programmatically inclined understand what is wrong.

ham1 commented on src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/TrustAllSocket.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public Socket createSocket(final String s, final int i, final InetAddress inetAddress, final int i1) throws IOException {
```
For readability could we please name the parameters the same as `SocketFactory`
```suggestion
    public Socket createSocket((String host, int port, InetAddress localHost, int localPort) throws IOException
```

wolfebrian2120 replied:
completed

bwolfehcm committed:
Updated TrustAllSocket overrided method parameters to match SocketFactory parameter names for better readability.

pmouawad commented:
Thanks @wolfebrian2120  for your contribution.
It is now merged but I made modifications. Can you checkout the currently building jenkins if it's successful:
https://builds.apache.org/job/JMeter-trunk/

Test and give us feedback.

Also, it would be better to make in GUI, Trust All only enabled if "Use Secure LDAP Protocol" is enabled as it is only used then.

Thanks

wolfebrian2120 commented:
Seems I am getting some stackOverFlow error when running it on my machine. I am not sure where it is originating. I didn't have any issues on my build. I even used my distribution to run against a prod environment.
Anyone have any thoughts on that. 

Thread Name: Thread Group 1-1
Sample Start: 2019-03-01 12:34:10 EST
Load time: 227
Connect Time: 0
Latency: 0
Size in bytes: 346
Sent bytes:0
Headers size in bytes: 0
Body size in bytes: 346
Sample Count: 1
Error Count: 1
Data type ("text"|"bin"|""): text
Response code: 800
Response message: javax.naming.CommunicationException: oem13c1.hcmlabs.net:1636 [Root exception is java.lang.StackOverflowError]


SampleResult fields:
ContentType: text/xml
DataEncoding: UTF-8

pmouawad commented:
Can you show jmeter.log content please ?
Thanks

wolfebrian2120 commented:
Here is the log file on trace, when I execute the 1 thread test.
[jmeter.log](https://github.com/apache/jmeter/files/2920782/jmeter.log)

pmouawad commented:
I see no trace of stackoverflow.
It looks like it happens in the ldap call client layer.
Does the build with your patch work ?
You think it is due to my changes ?

Thanks

wolfebrian2120 commented:
I think I know what went wrong. You removed the SSLSocketFactory object from the TrustAllSocketFactory. so you are only returning a Socket from a non-ssl socket factory. I think there is some behind the scenes magic in the SSLSocketFactory that does something the Socket that gets returned.

wolfebrian2120 commented:
Yeah you need to have some SocketFactory object to set the trust manager on. Then that factory has to create the Sockets. 
Here is the stackoverflow reference I used to do it.
https://stackoverflow.com/questions/4615163/how-to-accept-self-signed-certificates-for-jndi-ldap-connections/4615497
Here is another example I just found.
https://stackoverflow.com/questions/52600211/how-to-programmatically-disable-certificate-hostname-verification-in-java-ldap-j

pmouawad commented:
Yes.
I have just committed a fix. We already have such factory in mail protocol.
I moved it to refactor code between ldap and mail.
Sorry for that, this morning I wanted to cleanup and I missed this.

Regards

pmouawad commented:
Can you try:
https://builds.apache.org/job/JMeter-trunk/lastSuccessfulBuild/artifact/trunk/dist/

Thanks

wolfebrian2120 commented:
The last build (apache-jmeter-r1854620.zip) successfully passed my tests.
